knative-operator/pkg/controller/knativeserving/consoleclidownload/consoleclidownload.go

```diff
@@ -1,51 +1,310 @@
+	log.Info("Checking deployments")
+	for _, u := range manifest.Filter(mf.ByKind("Deployment"), mf.ByKind("Route")).Resources() {
+		switch u.GetKind() {
+		case "Deployment":
+			deployment := &appsv1.Deployment{}
+			err := api.Get(context.TODO(), client.ObjectKey{Namespace: u.GetNamespace(), Name: u.GetName()}, deployment)
+			if err != nil {
+				return err
+			}
+			for _, c := range deployment.Status.Conditions {
+				if c.Type == appsv1.DeploymentAvailable && c.Status != corev1.ConditionTrue {
+					return fmt.Errorf("Deployment %q/%q not ready", u.GetName(), u.GetNamespace())
+				}
+			}
+		case "Route":
+			route := &routev1.Route{}
+			if err := api.Get(context.TODO(), client.ObjectKey{Namespace: u.GetNamespace(), Name: u.GetName()}, route); err != nil {
+				return err
+			}
+			host := getCanonicalHost(route)
+			if host == "" {
+				return fmt.Errorf("Route %q/%q not ready yet", u.GetName(), u.GetNamespace())
+			}
+		}
+	}
+	return nil
+}
+
+// applyKnConsoleCLIDownload applies kn ConsoleCLIDownload by finding
+// kn download resource route URL and populating spec accordingly
+func applyKnConsoleCLIDownload(apiclient client.Client, instance *servingv1alpha1.KnativeServing) error {
+
+	log.Info("Installing kn ConsoleCLIDownload")
+	route := &routev1.Route{}
+	knConsole := &consolev1.ConsoleCLIDownload{}
+
+	// find the route first
+	if err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: instance.GetNamespace(), Name: knConsoleCLIDownloadDeployRoute}, route); err != nil {
+		return fmt.Errorf("failed to find kn ConsoleCLIDownload deployment route")
+	}
+
+	knRoute := getCanonicalHost(route)
+	knConsoleObj := populateKnConsoleCLIDownload(https(knRoute), instance)
+
+	// Check if kn ConsoleCLIDownload exists
+	err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: "", Name: knCLIDownload}, knConsole)
+	switch {
+	case apierrors.IsNotFound(err):
+		log.Info("Creating kn ConsoleCLIDownload")
+		if err := apiclient.Create(context.TODO(), knConsoleObj); err != nil {
+			return err
+		} else {
+			// handle upgrade scenario: In 1.6 we'd created kn ConsoleCLIDownload CO with
+			// name 'kn'. See https://github.com/openshift-knative/serverless-operator/blob/release-1.6/knative-operator/deploy/resources/console_cli_download_kn.yaml#L4
+			// lets delete the earlier CO if it exists
+			// TODO: Remove this post 1.7 release
+			knConsoleOld := &consolev1.ConsoleCLIDownload{}
+			if err := apiclient.Get(context.TODO(), client.ObjectKey{Namespace: "", Name: "kn"}, knConsoleOld); err == nil {
+				log.Info("Found old kn ConsoleCLIDownload CO 'kn', deleting it..")
+				if err := apiclient.Delete(context.TODO(), knConsoleOld); err != nil {
+					log.Info("failed to delete earlier kn ConsoleCLIDownload CO 'kn' %w", err)
+				}
+				log.Info("Deleted old kn ConsoleCLIDownload CO 'kn'")
+			}
+
+		}
+	case apierrors.IsAlreadyExists(err):
```

Comment:
This will never happen. A `Get` never returns a `IsAlreadyExists` error.

Instead, this function should be called `reconcileKnConsoleCLIDownload` and follow the layout of other reconcile functions. `reconcileConfigMap` for example, though this can be much less complex.

Reply:
Right.
Thanks for the ref, updated following `reconcileConfigMap`.